### PR TITLE
[FIRRTL][LowerClasses] Support lowering FIRRTL paths ops to OM

### PR DESCRIPTION
**Pull request by youngar**

### CODE REVIEW

File: test/Dialect/FIRRTL/lower-classes-errors.mlir [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// -----
+
+firrtl.circuit "PathDontTouchDeleted" {
+  firrtl.module @PathDontTouchDeleted() {}
+  firrtl.class @Class() {
+    // expected-error @+2 {{DontTouch target was deleted}}
+    // expected-error @+1 {{failed to legalize operation 'firrtl.path' that was explicitly marked illegal}}
+    %0 = firrtl.path dont_touch distinct[0]<>
+  }
+}
```
Comment (youngar) [inline test/Dialect/FIRRTL/lower-classes-errors.mlir:40]: Is there a way to not print the legalization error here?

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 2]
```diff
@@ -62,18 +83,144 @@ struct LowerClassesPass : public LowerClassesBase<LowerClassesPass> {
...
+    auto annotations = target.getAnnotations();
+    auto *op = target.getOp();
+    FModuleLike module;
+    annotations.removeAnnotations([&](Annotation anno) {
+      // If there has been an error, just skip this annotation.
+      if (error)
+        return false;
+
+      // We are looking for OMIR tracker annotations.
+      if (!anno.isClass("circt.tracker"))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:117]: Can this string constant live somewhere? Maybe with the other annotation class strings in that header?

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 3]
```diff
@@ -323,6 +471,61 @@ struct ListCreateOpConversion
...
+                  ConversionPatternRewriter &rewriter) const override {
+    auto *context = op->getContext();
+    auto pathType = om::PathType::get(context);
+    auto pathInfo = pathInfoTable.lookup(op.getTarget());
+
+    // If the target was optimized away, then replace the path operation with
+    // a deleted path.
+    if (!pathInfo) {
+      if (op.getTargetKind() == firrtl::TargetKind::DontTouch)
+        return emitError(op.getLoc(), "DontTouch target was deleted");
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:493]: Maybe `rewriter.notifyMatchFailure(...)`? (haven't tested, maybe that'll be combined/used with the error about legalization failing?)

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp (hunk 2 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:119]: will do this in a follow up

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 4]
```diff
@@ -62,18 +83,144 @@ struct LowerClassesPass : public LowerClassesBase<LowerClassesPass> {
...
+/// It builds a table which maps the OMIR trackker annotation IDs to the
+/// corresponding hierarchical paths. We use this table to convert FIRRTL path
+/// ops to OM. FIRRTL paths refer to their target using a target ID, while OM
+/// paths refer to their target using Hierarchical paths.
+static LogicalResult lowerPaths(CircuitOp circuit,
+                                PathInfoTable &pathInfoTable) {
+  SymbolTable symbolTable(circuit);
+  hw::InnerSymbolNamespaceCollection namespaces;
+  HierPathCache cache(circuit, symbolTable);
+  InstanceGraph instanceGraph(circuit);
```
[the remark this answers is not in the document]
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp]: Turns out I didn't need the instance graph at all.

**Pull request merged**